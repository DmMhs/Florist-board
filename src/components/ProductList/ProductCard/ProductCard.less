.ProductCard {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  border: 1px solid #eee;
  box-sizing: border-box;
  border-radius: 8px;
  margin: 4px 2px;
  font-weight: lighter;
  flex-basis: calc(20% - 6px);
  .info {
    cursor: pointer;
    font-size: 1.4rem;
    color: skyblue;
    position: absolute;
    top: 5px;
    right: 5px;
    z-index: 1;
    &:hover {
      color: rgb(0, 183, 255);
    }
  }
  .image {
    width: 165px;
    height: 165px;
    border-radius: 0%;
    border: 1px solid #eee;
    box-shadow: 1px 1px 5px #ccc;
    background-size: cover;
    background-position: center;
    margin-top: 2rem;
    margin-bottom: 1rem;
    .Slider {
      width: 100%;
      height: 100%;
      .LeftArrow {
        left: 0px;
      }
    }
  }
  .title {
    height: 2.5rem;
    font-size: 1.5rem;
    font-style: italic;
  }
  .price {
    height: 2.5rem;
    font-size: 1.2rem;
  }
  .action-panel {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    background-color: whitesmoke;
    width: 100%;
    height: 3rem;
    .action {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
      width: 100%;
      height: 2rem;
      padding: 5px;
    }

    i {
      cursor: pointer;
      font-size: 1.5rem;
      color: #333333;
      &.fa-cart-arrow-down:hover,
      &.active {
        color: rgb(98, 182, 103);
      }
      &.fa-heart:hover,
      &.fa-heart.active-action {
        color: tomato;
      }
    }
  }
}
@media screen and (max-width: 478px) {
  .ProductCard {
    flex-basis: calc(100% - 6px);
  }
}

@media screen and (min-width: 478.1px) and (max-width: 576px) {
  .ProductCard {
    flex-basis: calc(100% - 6px);
  }
}

@media screen and (min-width: 576.1px) and (max-width: 768px) {
  .ProductCard {
    flex-basis: calc(50% - 6px);
  }
}

@media screen and (min-width: 768.1px) and (max-width: 921px) {
  .ProductCard {
    flex-basis: calc(33.3% - 6px);
  }
}

@media screen and (min-width: 921.1px) and (max-width: 1200px) {
  .ProductCard {
    flex-basis: calc(25% - 6px);
    .image {
      width: 140px;
      height: 140px;
    }
  }
}
