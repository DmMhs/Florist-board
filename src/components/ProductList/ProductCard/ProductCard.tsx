import React, { Component } from 'react';

import './ProductCard.less';

class ProductCard extends Component<
  {
    title: string;
    imgSrc: string;
    price: number;
    currency: string;
  },
  any
> {
  constructor(props: any) {
    super(props);
    this.state = {
      image: 'assets/images/chamomile.jpg',
      title: 'Chamomile',
      price: 10,
      currency: 'usd',
      inCart: false,
      liked: false
    };
  }
  addToCartClickedHandler = (event: any) => {
    if (event.target.classList.contains('active-action')) {
      event.target.classList.remove('active-action');
      this.setState({
        inCart: false
      });
    } else {
      event.target.classList.add('active-action');
      this.setState({
        inCart: true
      });
    }
  };
  likeClickedHandler = (event: any) => {
    if (event.target.classList.contains('active-action')) {
      event.target.classList.remove('active-action');
      this.setState({
        liked: false
      });
    } else {
      event.target.classList.add('active-action');
      this.setState({
        liked: true
      });
    }
  };
  componentDidMount() {
    this.setState({
      image: this.props.imgSrc,
      title: this.props.title,
      price: this.props.price,
      currency: this.props.currency
    });
  }
  render() {
    return (
      <div className="ProductCard">
        <div
          className="image"
          style={{ background: `url(${this.state.image})` }}
        />
        <div className="title">{this.state.title}</div>
        <div className="price">
          {this.state.price}
          {this.state.currency}
        </div>
        <div className="action-panel">
          <i
            className="fas fa-cart-arrow-down"
            onClick={this.addToCartClickedHandler}
          />
          <i className="far fa-heart" onClick={this.likeClickedHandler} />
        </div>
      </div>
    );
  }
}

export default ProductCard;
