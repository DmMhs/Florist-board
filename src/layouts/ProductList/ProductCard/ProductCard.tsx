import React, { Component, RefObject } from 'react';
import { NavLink } from 'react-router-dom';

import { Slider } from '../../../components';
import { CartItem } from '../../../models/CartItem';
import { AppContext } from '../../../AppContext';
import { productsRef } from '../../../firebase';
import './ProductCard.less';
import { deleteProduct } from '../../../services/deleteProduct';

declare global {
  interface Window {
    fbAsyncInit: () => void;
  }
}

interface ProductCardState {
  isLiked: boolean;
  isLikedBy: string[];
}

class ProductCard extends Component<CartItem, ProductCardState> {
  public static getDerivedStateFromProps(
    props: CartItem,
    state: ProductCardState
  ) {
    return {
      isLiked: props.likedBy
    };
  }

  private likeButtonRef: RefObject<HTMLDivElement>;

  constructor(props: CartItem) {
    super(props);
    this.likeButtonRef = React.createRef();
    this.state = {
      isLiked: false,
      isLikedBy: []
    };
  }
  public componentDidMount() {
    if (this.state.isLiked === true) {
      this.likeButtonRef.current!.classList.add('active');
    }
  }

  private likeClickedHandler = async () => {
    this.likeButtonRef.current!.classList.toggle('active');
    await productsRef
      .child(this.props.id)
      .child('likedBy')
      .on('value', snapshot => {
        this.setState({
          isLiked: !this.state.isLiked,
          isLikedBy: snapshot!.val() === null ? [] : snapshot!.val()
        });
      });
    if (this.state.isLiked === true) {
      const newLikedByList = [...this.state.isLikedBy];
      const index = newLikedByList.findIndex((val: string) => {
        return val === this.context.state.userId;
      });
      newLikedByList.splice(index, 1);
      productsRef
        .child(this.props.id)
        .child('likedBy')
        .set(newLikedByList);
    } else {
      await productsRef.child(this.props.id).update({
        likedBy: [...this.state.isLikedBy, this.context.state.userId]
      });
    }
  };

  public render() {
    console.log(this.props);
    const {
      id,
      images,
      title,
      title_ua,
      price,
      currency,
      available,
      inCart
    } = this.props;

    const context = this.context;
    const labels = context.state.labels;
    const lang = context.state.lang;

    const actionIcon = inCart ? (
      <div>
        <div className="action">
          <i className="far fa-check-circle active" />
        </div>
      </div>
    ) : (
      <div>
        <div className="action">
          <i
            className="fas fa-cart-arrow-down"
            onClick={this.props.addToCart}
          />
        </div>
      </div>
    );

    const adminIcons =
      context.state.userRole === 'admin' ? (
        <div className="adminIcons">
          <NavLink to={`/admin/edit-product?id=${id}`}>
            <i className="far fa-edit" />
          </NavLink>
          <i
            className="far fa-trash-alt"
            onClick={deleteProduct.bind(this, id, title.toLowerCase())}
          />
        </div>
      ) : null;

    return (
      <AppContext.Consumer>
        {value =>
          value && (
            <div className="ProductCard">
              {adminIcons}
              <NavLink to={`/product-details/${id}`}>
                <i className="fas fa-info-circle info" />
              </NavLink>
              {value.state.userAuthenticated === true ? (
                <i
                  className="fas fa-heart like"
                  onClick={this.likeClickedHandler}
                  ref={this.likeButtonRef}
                />
              ) : null}
              <div className="image">
                <Slider images={images} auto={false} showControls={true} />
              </div>
              <div className="title">{lang === 'en' ? title : title_ua}</div>
              {available === true ? (
                <div className="price">
                  {price}
                  {currency}
                </div>
              ) : (
                <div className="price">
                  {labels[lang].pages.shop.notAvailable}
                </div>
              )}
              <div className="action-panel">
                {available ? actionIcon : null}{' '}
              </div>
            </div>
          )
        }
      </AppContext.Consumer>
    );
  }
}

ProductCard.contextType = AppContext;

export default ProductCard;
