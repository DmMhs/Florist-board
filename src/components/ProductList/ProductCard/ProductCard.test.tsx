import React from 'react';
import { shallow } from 'enzyme';
import {MemoryRouter} from 'react-router-dom';

import ProductCard from './ProductCard';

it('renders without crashing', () => {
  const wrapper = shallow(
    <MemoryRouter>
      <ProductCard
        title="some"
        images={['test', 'product']}
        price={10.5}
        currency="usd"
        available={false}
        key={0}
        id={'asfasf'}
        inCart={false}
      />
    </MemoryRouter>
  );
  expect(wrapper.find(ProductCard)).toMatchSnapshot()
});

it('Property "available" has impact on price displaying', () => {
  const p = {
    title: 'some',
    images: ['test', 'product'],
    price: 10.5,
    currency: 'usd',
    available: false,
    key: 0,
    id: 'asfasf',
    inCart: false,
    addToCart: () => {}
  };
  let wrapper = shallow(
    <ProductCard
      title="some"
      images={['test', 'product']}
      price={10.5}
      currency="usd"
      available={false}
      key={0}
      id={'asfasf'}
      inCart={false}
      addToCart={() => {}}
    />
  );
  expect(wrapper.find('.price').text()).toEqual('not available :(');
  wrapper = shallow(
    <ProductCard
      title="some"
      images={['test', 'product']}
      price={10.5}
      currency="usd"
      available={true}
      key={0}
      id={'asfasf'}
      inCart={false}
      addToCart={() => {}}
    />
  );
  expect(wrapper.find('.price').text()).toEqual('10.5usd');
});
