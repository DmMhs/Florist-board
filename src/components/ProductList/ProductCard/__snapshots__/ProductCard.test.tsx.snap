// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<ProductCard
  available={false}
  currency="usd"
  id="asfasf"
  images={
    Array [
      "test",
      "product",
    ]
  }
  inCart={false}
  key="0"
  price={10.5}
  title="some"
/>
`;
